libraries/cbor/src/macros.rs

```diff
@@ -12,6 +12,127 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use crate::values::{KeyType, Value};
+use alloc::collections::btree_map;
+use core::cmp::Ordering;
+use core::iter::Peekable;
+
+/// This macro generates code to extract multiple values from a `BTreeMap<KeyType, Value>` at once
+/// in an optimized manner, consuming the input map.
+///
+/// It takes as input a `BTreeMap` as well as a list of identifiers and keys, and generates code
+/// that assigns the corresponding values to new variables using the given identifiers. Each of
+/// these variables has type `Option<Value>`, to account for the case where keys aren't found.
+///
+/// **Important:** Keys passed to the `read_cbor_map!` macro **must be sorted** in increasing order.
+/// If not, the algorithm can yield incorrect results, such a assigning `None` to a variable even if
+/// the corresponding key existed in the map. **No runtime checks** are made for this in the
+/// `read_cbor_map!` macro, in order to avoid overhead at runtime. However, assertions that keys are
+/// sorted are added in `cfg(test)` mode, so that unit tests can verify ahead of time that the keys
+/// are indeed sorted. This macro is therefore **not suitable for dynamic keys** that can change at
+/// runtime.
+///
+/// Semantically, provided that the keys are sorted as specified above, the following two snippets
+/// of code are equivalent, but the `read_cbor_map!` version is more optimized, as it doesn't
+/// re-balance the `BTreeMap` for each key, contrary to the `BTreeMap::remove` operations.
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let map = alloc::collections::BTreeMap::new();
+/// read_cbor_map! {
+///     map,
+///     x @ 1,
+///     y @ 2,
+/// };
+/// # }
+/// ```
+///
+/// ```rust
+/// # extern crate alloc;
+/// # #[macro_use]
+/// # extern crate cbor;
+/// #
+/// # fn main() {
+/// #     let mut map = alloc::collections::BTreeMap::<cbor::KeyType, _>::new();
+/// let x: Option<cbor::Value> = map.remove(&cbor_unsigned!(1));
+/// let y: Option<cbor::Value> = map.remove(&cbor_unsigned!(2));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! read_cbor_map {
+    ( $map:expr, $( $variable:ident @ $key:expr, )+ ) => {
+        // A pre-requisite for this algorithm to work is that the keys to extract from the map are
+        // sorted.
+        #[cfg(test)]
+        test_ordered_keys!($( $key, )+);
+
+        use $crate::values::{IntoCborKey, Value};
+        use $crate::macros::read_cbor_map_peek_value;
+
+        let mut it = $map.into_iter().peekable();
+        $(
+        let $variable: Option<Value> = read_cbor_map_peek_value(&mut it, $key.into_cbor_key());
+        )+
+    };
+}
+
+/// This function is an internal detail of the `read_cbor_map!` macro, but has public visibility so
+/// that users of the macro can use it.
```

Comment:
What would be a usage scenario for this function?

Reply:
The function is only used inside of the macro. But because when using the macro the macro contents are expanded into some code, it's required that the functions introduced by this expansion are visible in the current scope. I'm not aware of a way to restrict that, but maybe @ia0 has an idea?

Reply:
I don't know any solution. What I do in those cases is to put `internal` in the name of the variable (function name, macro name, etc) and add `#[doc(hidden)]`.

```
#[doc(hidden)]
pub fn internal_read_cbor_map_peek_value(...) ...
// or
pub fn read_cbor_map_peek_value_internal(...) ...
```

Reply:
Ack. We don't currently generate any docs, but in any case the documentation for this function clearly states that it's an internal detail. I guess this sounds good then?

Reply:
Yes, the `#[doc(hidden)]` is only useful if the library is used by others (such that they are not tempted to use it). However, the `internal` in the name is a pattern useful even for us: a reviewer would see from the function name that some additional check has to be done (in a similar way to `unsafe` and `unchecked`). But I guess for this particular case, there's low chance someone would call it (by mistake or intentionally). This is mostly to give rationale of why those patterns can be useful.